packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts

```diff
@@ -275,6 +289,45 @@ export class AmplifyGraphqlApi extends Construct {
     outputStorageStrategy.addBackendOutputEntry(graphqlOutputKey, output);
   }
 
+  /**
+   * Gets the deployment information from the context.
+   * @returns the deployment identifier.
+   */
+  private getDeploymentIdentifier(): DeploymentIdentifier {
+    const deploymentType = this.node.tryGetContext(CDKContextKey.DEPLOYMENT_TYPE);
+    const expectedDeploymentTypes = [this.SANDBOX_DEPLOYMENT_TYPE, this.BRANCH_DEPLOYMENT_TYPE];
+
+    // Log a warning if the deployment type is neither sandbox nor branch.
+    if (!expectedDeploymentTypes.includes(deploymentType)) {
+      console.warn(
+        `${CDKContextKey.DEPLOYMENT_TYPE} is recommended to be one of (${expectedDeploymentTypes.join(', ')}).\n` +
+          'Consider setting the deployment type by running:\n' +
+          'cdk --context amplify-backend-type=sandbox\n' +
+          'or\n' +
+          'cdk --context amplify-backend-type=branch\n',
```

Comment:
I think this warning is going to cause more confusion than clarity. We don't intend for customers to interact directly with these context values so they aren't going to know what this mean. Also, saying "deployment type is recommended to be... consider setting..." is a lot of weasel words and immediately raises the question "what are the implications of setting or not setting?". If I saw this warning I would just be confused and probably ignore it

Reply:
We also already have [this mechanism](https://github.com/aws-amplify/amplify-backend/blob/d8b43d252c7d5553bf719dd889571807098adf13/packages/backend-data/src/factory.ts#L264-L272) for changing GSI update behavior in sandbox vs non-sandbox. Perhaps we could follow a similar pattern to inject necessary context into the data construct rather than the construct trying to fetch it.

Reply:
> We also already have [this mechanism](https://github.com/aws-amplify/amplify-backend/blob/d8b43d252c7d5553bf719dd889571807098adf13/packages/backend-data/src/factory.ts#L264-L272) for changing GSI update behavior in sandbox vs non-sandbox. Perhaps we could follow a similar pattern to inject necessary context into the data construct rather than the construct trying to fetch it?

We'd actually prefer to go the opposite direction, and use sandbox detection to drive the GSI update behavior, rather than make CLI flags privy to more of the transformer internals. We'd have to use aspects to hook onto a specific part of the target stack (e.g., the code you link has to know about the existence of `AmplifyDynamoDbTableWrapper`), and so the CLI now has to be tightly coupled to some part of the underlying stack. Context seems to me the right way for specifying environmental information, and to let the consuming stack do whatever it needs to.

Reply:
I think the "most correct" thing to do is to expose this behavior and the GSI behavior as a prop on the DataConstruct. Then the `backend-data` package can decide how to instantiate the construct based on the Amplify context it knows about and CDK consumers can do the same. This solution isn't eliminating tight coupling, it's just moving it. You are depending on the name and values of the CDK context set by `amplify-backend` without an explicit interface guarding it.

Reply:
Fair point about "moving the coupling without an interface."